### ISSUE TITLE
 Segmentation fault found 

### ISSUE THREAD

**Issue by FanWan** [JohnJocoo/AhoCorasick#2]
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffd71ab700 (LWP 6213)]
AhoCorasick<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >::prepareRecurcive(AhoCorasick<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >::State*, unsigned long, unsigned long) () at /home/opt/compiler/gcc-8.2/gcc-8.2/include/c++/8.2.0/bits/unique_ptr.h:342
342	      get() const noexcept

**Comment by JohnJocoo** [comment]
It is interesting, I don't know how creation of matcher can cause crash. @FanWan could you provide minimum code example that cause a crash? Or at least describe a context in which you use this matcher? If I can pinpoint the conditions which cause a crash, it should be easy to fix.

**Comment by FanWan** [comment]
Thanks for your reply。The crash occurred under multi-reader one-writer condition.

I will show you a brief description of my code.

Firstly,multi reading threads do search job as followings:

int Service:: ac_machine_search(std::vector<std::string> mentions, const std::string& text) {
 boost::shared_lock<boost::shared_mutex> lock(_mutex);
    auto matcher = this-> _ac_machine->matcher();
     // do something like matching
    return 0;
}


Then, a crontab writing thread, which runs every 10 minutes, do something like updating trie tree:

int Service::updated() {
    std::shared_ptr<ac_machine> new_ac_machine = std::make_shared<ac_machine>();

// do something to construct a new trie tree of new_ac_machine
……
……

boost::unique_lock<boost::shared_mutex> lock(_mutex);
// update pointer
this->_ac_machine = new_ac_machine;
}


PS: The member variable _ac_machine above is a shared pointer which points to trie tree of your implementation. Besides, i suspect there shoud be inappropriate operations of my code.

During performance tests, such as 100qps of reading threads and a crontab writing thread, crash occured.

Interestingly, when the crontab writing thread ran ony once(no more upating trie tree during tests), there was no crash.

**Comment by FanWan** [comment]
Shared read lock is used when searching while exclusive write lock is used when updating.  It seems that there are memory conflicts.

**Comment by JohnJocoo** [comment]
I am pretty confident now on what have caused a crash. When you get a matcher `this->_ac_machine->matcher()`, the new matcher object is being created and during creation it ensures that internal pattern matching tree is constructed inside of _ac_machine. So calling matcher() is actually a modifying (write) operation first time you call it (all following times it is read operation). From your example I see now how the library interface and naming can be confusing and I definitely should have spent more effort designing the library classes. But as an immediate fix you can do one of two things:

1 - Force internal pattern matching tree creation in update():
```
int Service::updated() {
  std::shared_ptr<ac_machine> new_ac_machine = std::make_shared<ac_machine>();
  // do something to construct a new trie tree of new_ac_machine
  ……
  // force internal pattern matching tree creation to prevent modifications in ac_machine_search()
  std::ignore = new_ac_machine->matcher(); // make sure this is not optimised away by compiler
  
  boost::unique_lock<boost::shared_mutex> lock(_mutex);
  // update pointer
  this->_ac_machine = new_ac_machine;
}
```

2 - Create matcher in update() and save it as class member:
```
int Service::updated() {
  std::shared_ptr<ac_machine> new_ac_machine = std::make_shared<ac_machine>();
  // do something to construct a new trie tree of new_ac_machine
  ……
  const auto new_ac_matcher = new_ac_machine->matcher();
  
  boost::unique_lock<boost::shared_mutex> lock(_mutex);
  // update pointers
  this->_ac_machine = new_ac_machine;
  this->_ac_matcher = std::make_shared<ac_machine::Matcher>(new_ac_matcher);
}

int Service:: ac_machine_search(std::vectorstd::string mentions, const std::string& text) {
  boost::shared_lock<boost::shared_mutex> lock(_mutex);
  auto matcher = *this->_ac_matcher;
  // do something like matching
}
```

Please tell me if you find that any of solutions fix the crash or not.

**Comment by FanWan** [comment]
thank you,i will try.

**Comment by FanWan** [comment]
It seems that the Solution #1  fix the crash, while the Solution #2 not. 
Continuous performance tests will be done to prove the Solution #1.
Thanks.

**Comment by JohnJocoo** [comment]
Closing this issue, seems like one of the suggestions fixed the crash.
